quinn-proto/src/connection/send_buffer.rs

```diff
@@ -1,30 +1,160 @@
 use std::{collections::VecDeque, ops::Range};
 
-use bytes::{Buf, Bytes};
+use bytes::{Buf, BufMut, Bytes};
 
 use crate::{VarInt, range_set::RangeSet};
 
 /// Buffer of outgoing retransmittable stream data
 #[derive(Default, Debug)]
 pub(super) struct SendBuffer {
-    /// Data queued by the application but not yet acknowledged. May or may not have been sent.
-    unacked_segments: VecDeque<Bytes>,
-    /// Total size of `unacked_segments`
-    unacked_len: usize,
-    /// The first offset that hasn't been written by the application, i.e. the offset past the end of `unacked`
-    offset: u64,
-    /// The first offset that hasn't been sent
+    /// Data queued by the application that has to be retained for resends.
+    ///
+    /// Only data up to the highest contiguous acknowledged offset can be discarded.
+    /// We could discard acknowledged in this buffer, but it would require a more
+    /// complex data structure. Instead, we track acknowledged ranges in `acks`.
     ///
-    /// Always lies in (offset - unacked.len())..offset
+    /// Data keeps track of the base offset of the buffered data.
+    data: SendBufferData,
+    /// The first offset that hasn't been sent even once
+    ///
+    /// Always lies in `data.range()`
     unsent: u64,
     /// Acknowledged ranges which couldn't be discarded yet as they don't include the earliest
     /// offset in `unacked`
+    ///
+    /// All ranges must be within `data.range().start..(data.range().end - unsent)`, since data
+    /// that has never been sent can't be acknowledged.
```

Comment:
What if we get an ack for a range we haven't sent? Is that possible?

Reply:
It is possible by a malicious peer, yes. There is nothing that the protocol can do to stop that. You don't want the peer to be able attack you by doing stuff like that. (not sure if this is attackable here, not checked out the context. only answering the direct question).

Reply:
It's not attackable, I just need to know what can happen so I can figure out what true invariants are.

We do need to make sure that acks get truncated before adding them to the acks RangeSet. Not sure if the old version does that, but the new one should.